### ISSUE TITLE
Tests of cacctoos is not oop

### ISSUE THREAD

**Issue by neonailol** [yegor256/cactoos#588]
Right now tests use alot of static function from matchers,
Maybe they should be more oop style, like let's take this example

```java
    @Test
    public void iteratesEmptyList() {
        final List<String> list = new LinkedList<>();
        MatcherAssert.assertThat(
            "Can't iterate a list",
            new And(
                new Mapped<String, Scalar<Boolean>>(
                    new FuncOf<>(list::add, () -> true), Collections.emptyList()
                )
            ),
            new ScalarHasValue<>(
                Matchers.allOf(
                    Matchers.equalTo(true),
                    new MatcherOf<>(
                        value -> {
                            return list.isEmpty();
                        }
                    )
                )
            )
        );
    }
``` 

in oop style this would be

```java
    @Test
    public void iteratesEmptyList() {
        final List<String> list = new LinkedList<>();
        MatcherAssert.assertThat(
            "Can't iterate a list",
            new And(
                new Mapped<String, Scalar<Boolean>>(
                    new FuncOf<>(list::add, () -> true), Collections.emptyList()
                )
            ),
            new ScalarHasValue<>(
                new AllOf<Boolean>(
                    new IterableOf<>(
                        new IsEqual<>(true),
                        new MatcherOf<>(
                            value -> {
                                return list.isEmpty();
                            }
                        )
                    )
                )
            )
        );
    }
```

or this is overkill?

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this issue

**Comment by yegor256** [comment]
@neonailol where will we get all those matchers?

**Comment by neonailol** [comment]
@yegor256 they already exists in `org.hamcrest.core`, all of them are public, and since this library has package for tests, if some matchers is missing, its implementations it should go there

**Comment by yegor256** [comment]
@neonailol sounds like a good idea, but we can't just replace all static calls with classes. Very soon somebody will add them back again. We have to create some validator, that will crash the build if `Matchers.*` is used. Any ideas?

**Comment by neonailol** [comment]
@yegor256 there is plugin for maven [forbidden-apis](https://github.com/policeman-tools/forbidden-apis)

Sample configuration: add this profile to pom.xml

```xml
<profile>
      <id>forbiddenapis</id>
      <build>
        <plugins>
          <plugin>
            <groupId>de.thetaphi</groupId>
            <artifactId>forbiddenapis</artifactId>
            <version>2.4.1</version>
            <configuration>
              <failOnUnsupportedJava>false</failOnUnsupportedJava>
              <signaturesFiles>
                <signaturesFile>./src/test/resources/org/cactoos/forbidden.txt</signaturesFile>
              </signaturesFiles>
            </configuration>
            <executions>
              <execution>
                <goals>
                  <goal>testCheck</goal>
                </goals>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
    </profile>
```

create file `./src/test/resources/org/cactoos/forbidden.txt` with content
```text
@defaultMessage Do not use static methods
org.hamcrest.Matchers
```

run maven goal `forbiddenapis:testCheck`, and you wold have something similar to this

```text
[ERROR] Forbidden class/interface use: org.hamcrest.Matchers [Do not use static methods]
[ERROR]   in org.cactoos.collection.BehavesAsCollection (BehavesAsCollection.java:64)
[ERROR] Forbidden class/interface use: org.hamcrest.Matchers [Do not use static methods]
[ERROR]   in org.cactoos.collection.BehavesAsCollection (BehavesAsCollection.java:64)
[ERROR] Forbidden class/interface use: org.hamcrest.Matchers [Do not use static methods]
[ERROR]   in org.cactoos.collection.BehavesAsCollection (BehavesAsCollection.java:68)
[ERROR] Forbidden class/interface use: org.hamcrest.Matchers [Do not use static methods]
```

plugin also includes some bundled signatures for jdk unsafe methods, but for this examaple i excluded them

**Comment by llorllale** [comment]
@neonailol btw, since `0.33` all matchers for cactoos were moved to new project `llorllale/cactoos-matchers`.

**Comment by llorllale** [comment]
@neonailol so let's start by forbidding those APIs and refactoring as necessary

**Comment by llorllale** [comment]
@0crat in

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/cactoos/issues/588#issuecomment-388578598))

@llorllale Job [#588](https://github.com/yegor256/cactoos/issues/588) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://www.zerocracy.com/policy.html#29): +15 point(s) just awarded to @neonailol[/z](https://www.0crat.com/u/neonailol)

**Comment by 0crat** [comment]
The score of @neonailol[/z](https://www.0crat.com/u/neonailol) -271 is too low and will be reset: +271 point(s) just awarded to @neonailol[/z](https://www.0crat.com/u/neonailol)

**Comment by 0crat** [comment]
The job [#588](https://github.com/yegor256/cactoos/issues/588) assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/C63314D6Z/1361331353); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html)

**Comment by victornoel** [comment]
@llorllale shouldn't the maven configuration go into `jcabi-parent` instead? or shouldn't we have a parent for cactoos projects maybe?

**Comment by llorllale** [comment]
@victornoel we'll use this project as our guinea pig. If the idea holds, maybe it'll spread to other projects as well

**Comment by victornoel** [comment]
@0crat waiting for #882

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@0caff0cf-8894-47fa-aeff-666b2b14aaac, thread: BkParallel-8
com.zerocracy.pm.Claims[112] java.lang.IllegalStateException: Duplicate claims are not allowed in [C63314D6Z](https://www.0crat.com/p/C63314D6Z), can't add this XML to 5 existing ones:
<?xml version="1.0" encoding="UTF-8"?>
<claim id="1391435999">
   <created>2018-05-19T14:35:38.348Z</created>
   <type>Register impediment</type>
   <token>github;yegor256/cactoos;588;390409117</token>
   <author>victornoel</author>
   <params>
      <param name="job">gh:yegor256/cactoos#588</param>
      <param name="repo">yegor256/cactoos</param>
      <param name="issue">588</param>
      <param name="comment">390409117</param>
   </params>
</claim>

```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm): Issue: yegor256/cactoos#588, Comment: 390409117

**Comment by 0crat** [comment]
> @0crat waiting for #882 ([here](https://github.com/yegor256/cactoos/issues/588#issuecomment-390409117))

@victornoel The impediment for [#588](https://github.com/yegor256/cactoos/issues/588) was registered successfully by @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
> @0crat waiting for #882 ([here](https://github.com/yegor256/cactoos/issues/588#issuecomment-390409117))

@victornoel Job [#588](https://github.com/yegor256/cactoos/issues/588) is already on hold

**Comment by victornoel** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/588#issuecomment-390412745))

@victornoel This is what I know about this job, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#588](https://github.com/yegor256/cactoos/issues/588) is [in scope](http://www.zerocracy.com/policy.html#14) for 7days
  * The role is `DEV`
  * The job is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel) for [3days](http://www.zerocracy.com/policy.html#8)
  * There is a monetary reward attached
  * @victornoel[/z](https://www.0crat.com/u/victornoel) will get 50% of their hourly rate for this job
  * The job has an [impediment](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 30 minutes/points
  * These users are banned and won't be assigned:
    * @neonailol[/z](https://www.0crat.com/u/neonailol): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%23588%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0pdd** [comment]
@neonailol 2 puzzles [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903) are still not solved.

**Comment by victornoel** [comment]
@neonailol #882 has been merged, this issue can now be closed

**Comment by 0crat** [comment]
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#588](https://github.com/yegor256/cactoos/issues/588) is now out of scope

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

**Comment by victornoel** [comment]
@llorllale 0crat didn't award me my +35 points :)

**Comment by llorllale** [comment]
@0crat pay 35min @victornoel

**Comment by 0crat** [comment]
> @0crat pay 35min @victornoel  ([here](https://github.com/yegor256/cactoos/issues/588#issuecomment-395095077))

@llorllale The project is under-funded, you can't do it now, see [§49](http://www.zerocracy.com/policy.html#49)

**Comment by victornoel** [comment]
@llorllale I think we can try again?

**Comment by victornoel** [comment]
@llorllale thanks

**Comment by 0pdd** [comment]
@neonailol 2 puzzles [#902](https://github.com/yegor256/cactoos/issues/902), [#951](https://github.com/yegor256/cactoos/issues/951) are still not solved; solved: [#903](https://github.com/yegor256/cactoos/issues/903).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#951](https://github.com/yegor256/cactoos/issues/951) is still not solved; solved: [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#1082](https://github.com/yegor256/cactoos/issues/1082) is still not solved; solved: [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#1119](https://github.com/yegor256/cactoos/issues/1119) is still not solved; solved: [#1082](https://github.com/yegor256/cactoos/issues/1082), [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#1166](https://github.com/yegor256/cactoos/issues/1166) is still not solved; solved: [#1082](https://github.com/yegor256/cactoos/issues/1082), [#1119](https://github.com/yegor256/cactoos/issues/1119), [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).

**Comment by 0pdd** [comment]
@neonailol 2 puzzles [#1222](https://github.com/yegor256/cactoos/issues/1222), [#1223](https://github.com/yegor256/cactoos/issues/1223) are still not solved; solved: [#1082](https://github.com/yegor256/cactoos/issues/1082), [#1119](https://github.com/yegor256/cactoos/issues/1119), [#1166](https://github.com/yegor256/cactoos/issues/1166), [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#1223](https://github.com/yegor256/cactoos/issues/1223) is still not solved; solved: [#1082](https://github.com/yegor256/cactoos/issues/1082), [#1119](https://github.com/yegor256/cactoos/issues/1119), [#1166](https://github.com/yegor256/cactoos/issues/1166), [#1222](https://github.com/yegor256/cactoos/issues/1222), [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#1434](https://github.com/yegor256/cactoos/issues/1434) is still not solved; solved: [#1082](https://github.com/yegor256/cactoos/issues/1082), [#1119](https://github.com/yegor256/cactoos/issues/1119), [#1166](https://github.com/yegor256/cactoos/issues/1166), [#1222](https://github.com/yegor256/cactoos/issues/1222), [#1223](https://github.com/yegor256/cactoos/issues/1223), [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#1583](https://github.com/yegor256/cactoos/issues/1583) is still not solved; solved: [#1082](https://github.com/yegor256/cactoos/issues/1082), [#1119](https://github.com/yegor256/cactoos/issues/1119), [#1166](https://github.com/yegor256/cactoos/issues/1166), [#1222](https://github.com/yegor256/cactoos/issues/1222), [#1223](https://github.com/yegor256/cactoos/issues/1223), [#1434](https://github.com/yegor256/cactoos/issues/1434), [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).

**Comment by 0pdd** [comment]
@neonailol the puzzle [#1910](https://github.com/yegor256/cactoos/issues/1910) is still not solved; solved: [#1082](https://github.com/yegor256/cactoos/issues/1082), [#1119](https://github.com/yegor256/cactoos/issues/1119), [#1166](https://github.com/yegor256/cactoos/issues/1166), [#1222](https://github.com/yegor256/cactoos/issues/1222), [#1223](https://github.com/yegor256/cactoos/issues/1223), [#1434](https://github.com/yegor256/cactoos/issues/1434), [#1583](https://github.com/yegor256/cactoos/issues/1583), [#902](https://github.com/yegor256/cactoos/issues/902), [#903](https://github.com/yegor256/cactoos/issues/903), [#951](https://github.com/yegor256/cactoos/issues/951).
